Fix for Wowza's RTMP stream

darkdarkdragon opened the pull request:
Wowza misses first video frame (with keyframe) and do not send
video until next keyframe, but sends audio. Probing buffer was not
large enough to accommodate all audio frames before video one
arrives.
Also stopping probing in case probe got both audio and video.
Parsing `onMetadata` frame to find out what types of streams will be
sent - this will allow to finish probing false - so it will improve latency
and reduce memory usage.

mtharp committed:
Fix overflow in calculating PCR

darkdarkdragon committed:
Fix for Wowza's RTMP stream

Wowza misses first video frame (with keyframe) and do not send
video until next keyframe, but sends audio. Probing buffer was not
large enough to accommodate all audio frames before video one
arrives.
Also stopping probing in case probe got both audio and video.
Parsing `onMetadata` frame to find out what types of streams will be
sent - this will allow to finish probing false - so it will improve latency
and reduce memory usage.

j0sh commented on format/flv/flv.go:
```diff
@@ -14,7 +14,7 @@ import (
 )
 
-var MaxProbePacketCount = 20
+var MaxProbePacketCount = 200
```
Is this still needed given that we're parsing onMetaData?

darkdarkdragon replied:
Yes - in my implementation at least. I'm not taking codec's info from metadata - only what streams present. For codecs info we still waiting for first audio and video packet.
Do you think it is reliable enough to take codec's info from metadata?

j0sh commented on format/rtmp/rtmp.go:
```diff
@@ -289,6 +289,27 @@ func (self *Conn) pollAVTag() (tag flvio.Tag, err error) {
+		case msgtypeidDataMsgAMF0:
+			if probing {
+				onMetaDataFound := false
+				for _, val := range self.datamsgvals {
```
Can we do this check within the `handleMsg` function so we aren't introducing a bool for `pollAVTag` ? The bool is only used during specific code paths. It's not ideal to have a control flow dependency based on what the function's caller is doing.

darkdarkdragon replied:
Don't want to put it into `handleMsg` - it works on every message, and checking for data inside messages require working with runtime type information - which I presume is not fast.
I've made commit with slight change - moved `onMetadata` check up in the chain.

j0sh replied:
Messages aren't sent very often (usually just a few at the beginning / end of a stream), and the runtime type lookup only needs to occur for the `onMetaData` message itself - which we have to do at some point anyway. The overall flow seems easier to reason about this way.

(We could also inline the onMetaData handling to avoid a second pass and the runtime lookup, but that's a more involved change.)

darkdarkdragon replied:
> and the runtime type lookup only needs to occur for the onMetaData message itself

To check if it is `onMetaData` message I need, for every message iterate over `[]interface{}`, check if its value is a string, and then check if it's equal `onMetaData`.

Also, handling onMetatData in `handleMsg` would require either set properties of `self.prober` directly from `handleMsg` - which looks really wrong, or to add new data structure and new field to `Conn` object to pass parsed `onMedata` object from `handleMsg` up to `probe` func, where it is really needed.

j0sh reviewed (commented):
Very nice! Just a couple notes.

darkdarkdragon committed:
Moved `onMetatdata` check up

mk-livepeer reviewed (commented):
How does this affect the segments during playback?   Are we chopping off some audio?

darkdarkdragon commented:
@mk-livepeer This would affect segmentation at all - It doesn't change anything in stream, it just do less caching at start of stream. FFmpeg will start segmentation at the next key frame anyway, this code has nothing to do with it.

darkdarkdragon committed:
Add corner case for atoms. When size is 1, the real size is the 8
bytes after the header tag

darkdarkdragon committed:
flv: ignore AAC SEQHDR with no data, which ffmpeg sometimes emits

darkdarkdragon committed:
Fix tsio.go

darkdarkdragon committed:
for Ts demuxer return all streams - or else indexes will not match with packet's indexes

j0sh commented:
> flv: ignore AAC SEQHDR with no data, which ffmpeg sometimes emits

For what it's worth: "sometimes emits" is specifically when converting from mpegts to FLV / RTMP. This can be mitigated on the sender side with the `-flvflags aac_seq_header_detect` option, like so:

```
ffmpeg -re -i test/in/testharness.m3u8 -c:a copy -c:v copy -f flv -flvflags aac_seq_header_detect rtmp://localhost
```

darkdarkdragon committed:
Define and return specific errors when info about codecs not found

darkdarkdragon committed:
TS muxer - do not add extra second on packet write

darkdarkdragon committed:
Add option for TS demuxer to parser files with arbitrary number of streams

darkdarkdragon committed:
MP4: Read and write Media Header 'mdhd' atom version 1

darkdarkdragon committed:
Read and write 'co64' atom

darkdarkdragon committed:
Round timestamp in the RTMP stream up - match ffmpeg's behaviour

darkdarkdragon committed:
formatting

darkdarkdragon committed:
improve bits_test

darkdarkdragon committed:
parse NALUs with RBSP
ParseNalu5 function
parse FPS out of SPS

darkdarkdragon committed:
Add TimeScale, TimeB, CompositionTimeB fields to the packet
Fill them for mp4

darkdarkdragon committed:
tools used in debugging

darkdarkdragon committed:
format files - no code changes

darkdarkdragon committed:
fix tests

darkdarkdragon committed:
hide CGO usage behind build tag

darkdarkdragon committed:
formatting

darkdarkdragon committed:
Rename TimeB to TimeTS
Add TimeScale to CodedData interface
Correctly mux and demux TimeTS for mp4 and ts

darkdarkdragon committed:
mp4 demuxer - correctly demux last packet

darkdarkdragon committed:
Add tests for .ts and .mp4

darkdarkdragon committed:
Fix overflow in calculating PCR

darkdarkdragon committed:
Fix overflow in calculating PCR

darkdarkdragon committed:
Fix TimeScaleTimeToTs

darkdarkdragon committed:
Fix TS muxer - write DTS for video stream even if it is zero

darkdarkdragon committed:
for TrackFragDecodeTime atom just read raw value

darkdarkdragon committed:
avutil.OpenRC for reading from memory
